OpenRA.Mods.Common/Traits/Buildings/ProductionAirdrop.cs

```diff
@@ -112,11 +103,142 @@ public override bool Produce(Actor self, ActorInfo producee, string productionTy
 					Game.Sound.PlayNotification(self.World.Map.Rules, self.Owner, "Speech", info.ReadyAudio, self.Owner.Faction.InternalName);
 				}));
 
-				actor.QueueActivity(new FlyOffMap(actor, Target.FromCell(w, endPos)));
-				actor.QueueActivity(new RemoveSelf());
+				actor.QueueActivity(new CallFunc(() =>
+				{
+					var exitPosition = deliveryPath.CalculateExitPosition(actor);
+					actor.QueueActivity(new FlyOffMap(actor, Target.FromCell(w, exitPosition)));
+					actor.QueueActivity(new RemoveSelf());
+				}));
 			});
 
 			return true;
 		}
+
+		DeliveryActorPathInfo GetDeliveryActorPathInfo(Actor self)
+		{
+			var owner = self.Owner;
+			var map = owner.World.Map;
+			var bounds = map.Bounds;
+			var mpStart = owner.World.WorldActor.TraitOrDefault<MPStartLocations>();
+
+			var startPos = CPos.Zero;
+			var spawnFacing = 0;
+
+			switch (info.EntryType)
+			{
+				case EntryType.Fixed:
+					// Start a fixed distance away: the width of the map.
+					// This makes the production timing independent of spawnpoint
+					var loc = self.Location.ToMPos(map);
+					startPos = new MPos(loc.U + map.Bounds.Width, loc.V).ToCPos(map);
+
+					spawnFacing = GetSpawnFacing(self.Location, startPos);
+					break;
+
+				case EntryType.PlayerSpawnClosestEdge:
+					if (mpStart == null)
+						break;
+
+					var spawn = mpStart.Start[owner];
+					var center = new MPos(bounds.Left + bounds.Width / 2, bounds.Top + bounds.Height / 2).ToCPos(map);
+					var spawnVec = spawn - center;
+					startPos = spawn + spawnVec * (Exts.ISqrt((bounds.Height * bounds.Height + bounds.Width * bounds.Width) / (4 * spawnVec.LengthSquared)));
+
+					spawnFacing = GetSpawnFacing(self.Location, startPos);
+					break;
+
+				case EntryType.DropSiteClosestEdge:
+					startPos = self.World.Map.ChooseClosestEdgeCell(self.Location);
+
+					spawnFacing = GetSpawnFacing(self.Location, startPos);
+
+					bounds = self.World.Map.Bounds;
+					if (info.SpawnOffset.X != 0 && startPos.X != bounds.X && startPos.X != bounds.X + bounds.Width)
+						startPos += new CVec(info.SpawnOffset.X, 0);
```

Comment:
What is the purpose of `SpawnOffset` and `ExitOffset`? Shouldn't they depend on the approach direction as well? And why do they only apply to the 'closest edge' case?